# An analysis of used car listing from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset is taken from here [kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data) , though it was a bit curated and modified by Dataquest. 

The data dictionary provided by the author is as follows:
- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

We will analyze this data, starting with cleaning it up


In [2]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

We can see that there are 5 columns with some null values, let's check the percentage of the nulls.

In [5]:
total_nulls_column = {}

for col in autos.columns:
    num_nulls = autos[col].isna().sum()
    total_nulls_column[col] = num_nulls / 500
    
sorted_vals = sorted(total_nulls_column.items(), key = lambda kv: kv[1], reverse=True)
for pair in sorted_vals:
    if pair[1] > 0:
        print('Column {} has {}% nulls'.format(pair[0], pair[1]))

Column notRepairedDamage has 19.658% nulls
Column vehicleType has 10.19% nulls
Column fuelType has 8.964% nulls
Column model has 5.516% nulls
Column gearbox has 5.36% nulls


In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [7]:
print(autos.columns)
len(autos.columns)


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


20

# Dataframe observations

- There are 50000 entries. 
- The number of columns is 20, with the CamelCase naming convention 
- Most of the data types are strings 
- There are null values, more specific 5 columns have them, with the column ```notRepairedDamage``` having the most, almost 20%

Let's clean up the columns names by converting them from camelcase to snakecase. 

We will create a function that takes a string in camecase and returns a snakecase one, using regex.

We do the conversion in order to have a more descriptive name. 


# Rename the columns 

In [8]:
import re

def to_snakecase(camelcase):
    c = re.compile('((?<=[a-z0-9])[A-Z]|(?!^)[A-Z](?=[a-z]))')
    snakecase = c.sub(r'_\1', camelcase).lower()
    return snakecase

In [9]:
columns = autos.columns
new_columns = []
for col in columns:
    new_columns.append(to_snakecase(col))

autos.columns = new_columns

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Actually some of the columns names can be improves a little, let's do that 

In [11]:
autos.rename({'year_of_registration':'registration_year'}, axis=1, inplace=True)
autos.rename({'month_of_registration':'registration_month'}, axis=1, inplace=True)
autos.rename({'not_repaired_damage':'unrepaired_damage'}, axis=1, inplace=True)

In [12]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Look over statistics to see what else needs cleaning up

Next step would be to check the statistics and see if there is any more cleanup that can be done

In [13]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

# Remove same value columns

- Looks like we can immediatly drop ```nr_of_pictures``` column, it's zero everywhere

In [14]:
autos = autos.drop(['nr_of_pictures'], axis=1)

- Some columns have only 2 values, we will look into those.

In [15]:
for col in ['seller', 'offer_type', 'abtest', 'gearbox', 'unrepaired_damage']:
    print('Info for column "{}"'.format(col))
    print(autos[col].value_counts())
    print('----------------------------------')

Info for column "seller"
privat        49999
gewerblich        1
Name: seller, dtype: int64
----------------------------------
Info for column "offer_type"
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
----------------------------------
Info for column "abtest"
test       25756
control    24244
Name: abtest, dtype: int64
----------------------------------
Info for column "gearbox"
manuell      36993
automatik    10327
Name: gearbox, dtype: int64
----------------------------------
Info for column "unrepaired_damage"
nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64
----------------------------------


```seller``` and ```offer_type``` columns have 99.999% same value, so it should be safe to drop them as well

In [16]:
autos = autos.drop(['seller', 'offer_type'], axis=1)

# Clean and analyze ['price', 'odometer'] columns

```price``` and ```odometer``` can be cleaned by removing the $ and respectivly Km and by converting them to numeric. It would make sense it we remove Km to rename the column ```odometer_km```

In [17]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [18]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage         date_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

With the ```price``` and ```odomoter_km``` as numbers, we can properly look into their statistics anc check for iregularities.

In [19]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Most of the cars are listed with high mileage.

In [20]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [21]:
autos['price'].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [22]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Max value its very high 100 millions, skewing a lot the rest of the results. Let's look more into it

In [23]:
autos['price'].value_counts()

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64

There are 1421 cars with a zero value, this is about 2% of all cars. We could probably ignore this.

Let's check the top high prices.

In [24]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

Seems that the prices go higher and higher, with a big leap from 350000 to 999990. It's unlikely that the cars on ebay would sell for that amount, a proper bid house would be the place to go for that. 

It make sense to remove then the cars with values over 350000. As for the lower threshold is should be enough to remove the cars with the selling price zero. It is relativly common on ebay to start a bid from 1 , so we can leave everything that start from 1 and up to 350 000. 

In [25]:
autos = autos[autos['price'].between(1, 351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Analyze dates columns

There are 5 columns for the dates. It looks like some were created by the crawler, while others are from ebay. 

In [26]:
autos[['date_crawled', 'last_seen','date_created','registration_month', 'registration_year']].dtypes

date_crawled          object
last_seen             object
date_created          object
registration_month     int64
registration_year      int64
dtype: object

Three of them are strings, while the other two are numbers already.

Let's have a look over the string columns.

In [27]:
autos[['date_crawled', 'last_seen','date_created']].head()

date_crawled            last_seen         date_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

We have a pattern for the strings describing the dates. The first 10 characters represent the dates, followed by a whitespace, with the rest of the 8 characters describing the time of day.

We can check the distribution of the dates by extracting the first 10 characters.

In [28]:
print(autos['date_crawled'].str[:10].value_counts())
autos['date_crawled'].str[:10].value_counts().shape

2016-04-03    1875
2016-03-20    1840
2016-03-21    1815
2016-03-12    1793
2016-03-14    1775
2016-04-04    1772
2016-03-07    1749
2016-04-02    1723
2016-03-28    1693
2016-03-19    1689
2016-03-15    1665
2016-03-29    1656
2016-04-01    1636
2016-03-30    1636
2016-03-08    1617
2016-03-09    1607
2016-03-22    1602
2016-03-11    1582
2016-03-23    1565
2016-03-26    1564
2016-03-10    1563
2016-03-31    1546
2016-03-17    1536
2016-03-25    1535
2016-03-27    1510
2016-03-16    1438
2016-03-24    1425
2016-03-05    1230
2016-03-13     761
2016-03-06     682
2016-04-05     636
2016-03-18     627
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64


(34,)

Let's check for null values and also get the count as percentage

In [29]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
autos['date_crawled'].str[:10].value_counts().shape

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


(34,)

In [30]:
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_values())
autos['date_crawled'].str[:10].value_counts().shape

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64


(34,)

Looks like the data was retrieved by crawling for about a month, between march and april 2016. The data retrieved is reltavily even distributed.

In [31]:
print(autos['date_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
autos['date_created'].str[:10].value_counts().shape

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

(76,)

The ad creation date varies a lot , with the listing distribution being less uniform.

In [32]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
autos['last_seen'].str[:10].value_counts().shape

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


(34,)

The dates seem to be similar with the crawling ones, which makes sense, cars were sold. 

As for the listings distribution, last 3 days show a big jump , lots of sales maybe? we can look into those dates and see the date difference.

In [33]:
autos.loc[autos['date_crawled'].str[:10]=='2016-04-06', ['date_crawled','last_seen']]

date_crawled            last_seen
213    2016-04-06 15:06:36  2016-04-06 15:25:17
229    2016-04-06 14:06:18  2016-04-06 14:06:18
624    2016-04-06 18:06:33  2016-04-06 18:25:18
752    2016-04-06 20:06:38  2016-04-06 20:25:20
1060   2016-04-06 12:36:34  2016-04-06 12:36:34
1703   2016-04-06 07:06:33  2016-04-06 08:25:16
1853   2016-04-06 12:25:21  2016-04-06 12:25:21
2024   2016-04-06 10:06:34  2016-04-06 10:25:19
2681   2016-04-06 19:06:23  2016-04-06 19:06:23
2746   2016-04-06 19:25:25  2016-04-06 19:25:25
3307   2016-04-06 02:36:16  2016-04-06 02:36:16
3864   2016-04-06 20:25:23  2016-04-06 20:25:23
4299   2016-04-06 21:25:35  2016-04-06 21:25:35
4420   2016-04-06 15:25:21  2016-04-06 15:25:21
4456   2016-04-06 14:25:24  2016-04-06 14:25:24
4613   2016-04-06 06:36:22  2016-04-06 06:36:22
5208   2016-04-06 16:06:33  2016-04-06 16:25:16
6192   2016-04-06 12:06:23  2016-04-06 12:25:17
6545   2016-04-06 15:25:25  2016-04-06 15:25:25
6755   2016-04-06 12:36:22  2016-04-06 12:36:22
7129   2016-04-06 12:36:25  2016-04-06 12:36:25
7180   2016-04-06 19:25:22  2016-04-06 19:25:22
7181   2016-04-06 18:25:35  2016-04-06 18:25:35
7852   2016-04-06 08:36:21  2016-04-06 08:36:21
8011   2016-04-06 06:36:17  2016-04-06 06:36:17
8149   2016-04-06 02:36:32  2016-04-06 02:36:32
8298   2016-04-06 03:36:37  2016-04-06 04:36:18
8402   2016-04-06 06:36:22  2016-04-06 06:36:22
8470   2016-04-06 17:06:33  2016-04-06 17:25:19
8776   2016-04-06 02:36:21  2016-04-06 02:36:21
...                    ...                  ...
39091  2016-04-06 12:06:23  2016-04-06 12:25:16
39202  2016-04-06 15:25:38  2016-04-06 15:25:38
39333  2016-04-06 23:25:21  2016-04-06 23:25:21
39995  2016-04-06 15:36:22  2016-04-06 15:36:22
40221  2016-04-06 00:06:21  2016-04-06 00:25:16
40635  2016-04-06 17:06:35  2016-04-06 17:25:22
41219  2016-04-06 11:36:23  2016-04-06 11:36:23
41299  2016-04-06 15:36:32  2016-04-06 15:36:32
41312  2016-04-06 23:36:22  2016-04-06 23:36:22
41353  2016-04-06 16:25:34  2016-04-06 16:25:34
41452  2016-04-06 20:06:24  2016-04-06 20:06:24
41689  2016-04-06 23:06:34  2016-04-06 23:25:19
42133  2016-04-06 21:25:35  2016-04-06 21:25:35
42349  2016-04-06 00:06:37  2016-04-06 02:25:17
42982  2016-04-06 23:06:32  2016-04-06 23:25:16
43032  2016-04-06 19:25:38  2016-04-06 19:25:38
43241  2016-04-06 13:06:38  2016-04-06 13:25:20
43453  2016-04-06 11:06:25  2016-04-06 11:25:18
44015  2016-04-06 21:06:15  2016-04-06 21:06:15
44460  2016-04-06 01:36:18  2016-04-06 01:36:18
45044  2016-04-06 20:36:19  2016-04-06 20:36:19
45197  2016-04-06 08:06:37  2016-04-06 09:25:16
45324  2016-04-06 18:36:35  2016-04-06 18:36:35
45393  2016-04-06 12:36:20  2016-04-06 12:36:20
46160  2016-04-06 07:36:18  2016-04-06 07:36:18
46348  2016-04-06 10:06:19  2016-04-06 10:06:19
46973  2016-04-06 20:25:39  2016-04-07 14:29:12
47529  2016-04-06 16:06:17  2016-04-06 16:06:17
48190  2016-04-06 11:36:17  2016-04-06 11:36:17
48488  2016-04-06 15:06:18  2016-04-06 15:06:18

[154 rows x 2 columns]

Doesnt make a lot of sense why such a big jump starting with 3rd day to last of crawling, unless data was crawled unevenly somewhat or the cleanup of this subset of dataset ( it's not the original ) was modified.

# Analyze `registration_year`
Let's check the `registration_year` too

In [34]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Looks like there are a few eronous dates too, unless someone managed to build a car in year 1000 or bring one from the future.

In [35]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

Safe to say we can cleanup cars from 2800 and up and those under 1910. 

The upper date should actually be 2016, since we have only crawled data from 2 months of 2016.

In [36]:
autos = autos[autos['registration_year'].between(1900, 2016)]
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [37]:
autos['registration_year'].value_counts(normalize=True).sort_values(ascending=False).head(30)


2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
Name: registration_year, dtype: float64

Most of the cars are registered starting with 1994. We can see the exact percent under, 94% .

In [38]:
(autos["registration_year"].between(1994,2016)).sum() / autos.shape[0]

0.940489706732932

# Analyze `brand`

Let's have a look over the `brand` column and see what can we aggregate over. Let's start by having a look at the unique values.

In [39]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [40]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We can see the top 4 make out for around 50% of all brands ( all german brands ).

We could trim a bit the data by removing the brands with under 5% reprezentation

In [41]:
brands_counts = autos['brand'].value_counts(normalize=True)
brands_name = brands_counts[brands_counts > 0.05].index
print(brands_name)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


We have only 6 brands with more than 5% representation in our dataset. Let's looke over the average price per brand

In [42]:
price_average = {}
for brand in brands_name:
    avg = autos.loc[autos['brand'] == brand, 'price'].mean()
    price_average[brand] = int(avg)

sorted_values = sorted(price_average.items(), key= lambda kv: kv[1], reverse=True)
for pair in sorted_values:
    print('{} brand average sell price is: {}'.format(pair[0], pair[1]))
    

audi brand average sell price is: 9336
mercedes_benz brand average sell price is: 8628
bmw brand average sell price is: 8332
volkswagen brand average sell price is: 5402
ford brand average sell price is: 3749
opel brand average sell price is: 2975


Audi has the highest prices on average, with Opel being at the bottom .

Top 3 brands - Audi, Mercedes and BMW - have close average prices, followed by Volswagen in the middle and the last two again being close , Ford and Opel.

Let's check the mileage and see if there is any corelation between the price and mileage.

In [43]:
mileage_average = {}
for brand in brands_name:
    avg = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    mileage_average[brand] = int(avg)

sorted_values = sorted(mileage_average.items(), key= lambda kv: kv[1], reverse=True)
for pair in sorted_values:
    print('{} brand average mileage: {}'.format(pair[0], pair[1]))

bmw brand average mileage: 132572
mercedes_benz brand average mileage: 130788
opel brand average mileage: 129310
audi brand average mileage: 129157
volkswagen brand average mileage: 128707
ford brand average mileage: 124266


We can see already that the mileage is relativly close between the whole 6 brands

Let's combine these two statistics into one dataframe.

We start by making the two dictionaries into series and after combining them into a dataframe.

In [44]:
brands_prices = pd.Series(price_average)
brands_mileage = pd.Series(mileage_average)


mileage_price = pd.DataFrame(brands_prices, columns=['mean_price'])
mileage_price['mean_mileage'] = brands_mileage

mileage_price.sort_values(by='mean_price')


mean_price  mean_mileage
opel                 2975        129310
ford                 3749        124266
volkswagen           5402        128707
bmw                  8332        132572
mercedes_benz        8628        130788
audi                 9336        129157

By analyzing the `mileage_price` dataframe its quite visible there is no corelation between the mean mileage and the mean price.

The difference in prices is probably related more to the quality of the brand.

Let's go a bit deeper and split the mileage into groups and see if there is any corelation between that and average prices.

In [45]:
mileage_groups = autos['odometer_km'].value_counts().index
avg_price_mileage = {}

for miles in mileage_groups:
    avg = autos.loc[autos['odometer_km'] == miles, 'price'].mean()
    avg_price_mileage[miles] = avg
    
pd.Series(avg_price_mileage).sort_values(ascending=False)

10000     20550.867220
20000     18448.477089
30000     16608.836842
40000     15499.568381
50000     13812.173212
60000     12385.004433
70000     10927.182814
80000      9721.947636
5000       8873.515924
90000      8465.025105
100000     8132.697279
125000     6214.022030
150000     3767.927107
dtype: float64

As expected, less mileage cars have higher prices on average, with one exceptions. Cars with 5000 km would be expected to have the highest price, yet they are slighly under middle. 

# Analyze 5k mileage cars

Normally these cars should have the highest price on average, yet they are a bit under middle . Let's check why.


In [46]:
autos['odometer_km'].value_counts()

150000    30085
125000     4857
100000     2058
90000      1673
80000      1375
70000      1187
60000      1128
50000       993
40000       797
5000        785
30000       760
20000       742
10000       241
Name: odometer_km, dtype: int64

They seem to have a similar representation as some other medium mileage cars so this is not the reason for the low price on average. 

Let's check the brands of those with 5000 km.

In [53]:
fivek_brands = autos.loc[autos['odometer_km'] == 5000, 'brand']
fivek_brands.value_counts(normalize=True).head(10)

volkswagen        0.194904
opel              0.127389
bmw               0.098089
mercedes_benz     0.071338
ford              0.068790
sonstige_autos    0.061146
renault           0.050955
audi              0.043312
fiat              0.034395
porsche           0.019108
Name: brand, dtype: float64

Let's also check the registration year for these cars

In [59]:
fivek_reg_year_brands = autos.loc[autos['odometer_km'] == 5000, ['brand', 'registration_year', 'price']]
fivek_reg_year_brands.sort_values(by='registration_year')

brand  registration_year   price
3679   sonstige_autos               1910       1
45157         trabant               1910      15
28693         renault               1910     599
21416            ford               1927   16500
22101             bmw               1929   11500
2221             opel               1934    3350
21421            ford               1937    7000
23804   mercedes_benz               1937    4800
26103  sonstige_autos               1938    8500
24855  sonstige_autos               1939   26600
14020      alfa_romeo               1950   10000
1171             seat               1950       2
20471  sonstige_autos               1959    5500
8215    mercedes_benz               1960   50000
14070  sonstige_autos               1960    7500
23457            opel               1960    3500
4907        chevrolet               1960     100
49722      alfa_romeo               1960     200
21250      volkswagen               1961    8200
15934         porsche               1963   99000
12647            ford               1964    2800
37164         trabant               1964     450
15281         trabant               1964     500
19993            ford               1965   25999
42676            ford               1965   33900
28227  sonstige_autos               1966   25500
3796            volvo               1966    7777
21399            ford               1966    2890
5998             ford               1967   52000
12971             bmw               1968    4800
...               ...                ...     ...
24586            opel               2016   14900
43623      mitsubishi               2016       1
26877      volkswagen               2016    1600
25509            seat               2016   14290
26631         peugeot               2016    2800
26236         citroen               2016     100
19802            opel               2016   22900
45038       chevrolet               2016    6400
30587            opel               2016      17
31054            mini               2016       9
12607          nissan               2016   10490
46782            audi               2016   49900
46447            ford               2016     350
12682         porsche               2016  265000
39976            audi               2016    3400
13604            audi               2016   18500
33744            audi               2016   34900
33584         renault               2016    1000
12322         renault               2016   20900
32981      volkswagen               2016   64280
16552         citroen               2016   10990
16803          suzuki               2016    1490
32143            opel               2016     750
17837            opel               2016   14850
45422      volkswagen               2016   29700
45279      volkswagen               2016   25900
31836            seat               2016   12800
18264         peugeot               2016    9300
40269            mini               2016   23540
1956       volkswagen               2016    2700

[785 rows x 3 columns]

Seems there are cars with recent registration years yet with low price. Might be we are dealing with cars who were new yet involved in accidents and sold for less.

In [63]:
fivek_unrepaired = autos.loc[autos['odometer_km'] == 5000, 'unrepaired_damage']
fivek_unrepaired.value_counts(normalize
                              =True)

nein    0.805936
ja      0.194064
Name: unrepaired_damage, dtype: float64

Almost 20% are damaged, this could be a reason why. Let's chech this for a few of the other low mileage groups.

In [65]:
tenk_unrepaired = autos.loc[autos['odometer_km'] == 10000, 'unrepaired_damage']
tenk_unrepaired.value_counts(normalize=True)

nein    0.961165
ja      0.038835
Name: unrepaired_damage, dtype: float64

In [66]:
twentyk_unrepaired = autos.loc[autos['odometer_km'] == 20000, 'unrepaired_damage']
twentyk_unrepaired.value_counts(normalize=True)

nein    0.964992
ja      0.035008
Name: unrepaired_damage, dtype: float64

In [67]:
thirtyk_unrepaired = autos.loc[autos['odometer_km'] == 30000, 'unrepaired_damage']
thirtyk_unrepaired.value_counts(normalize=True)

nein    0.959596
ja      0.040404
Name: unrepaired_damage, dtype: float64

Looks like the reason 5k mileage cars are not the most expensive on average is because a high percentage are sold damaged, around 20%. 10k and 20k mileage have around 3% damaged, 30k around 4%. 

# Check the price difference between damaged and non-damaged cars on average

In [76]:
avg_price_undamaged_autos = autos.loc[autos['unrepaired_damage']=='nein', 'price'].mean()
print('Undamaged autos average price: {:.2f}'.format(avg_price_undamaged_autos))
avg_price_damaged_autos = autos.loc[autos['unrepaired_damage']=='ja', 'price'].mean()
print('Damaged autos average price: {:.2f}'.format(avg_price_damaged_autos))


Undamaged autos average price: 7164.03
Damaged autos average price: 2241.15


The difference is quite big.

Let's have a look at how the 5k mileage cars price differs

In [77]:
fivek_avg_price_undamaged_autos = autos.loc[(autos['unrepaired_damage']=='nein') & (autos['odometer_km']==5000), 'price'].mean()
print('5k mileage undamaged autos average price: {:.2f}'.format(fivek_avg_price_undamaged_autos))
fivek_avg_price_damaged_autos = autos.loc[(autos['unrepaired_damage']=='ja') & (autos['odometer_km']==5000), 'price'].mean()
print('5k mileage Damaged autos average price: {:.2f}'.format(fivek_avg_price_damaged_autos))

5k mileage undamaged autos average price: 16277.10
5k mileage Damaged autos average price: 1699.74


As expected , the difference amplifies here, from around 3.3 timescheaper to around 10 times cheaper compared to undamaged.